### ISSUE TITLE
Bug: [prefer-readonly-parameter-types] Does not work with ReadonlyMap or ReadonlySet

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#8013]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&showAST=ts&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBDATljBPALkQGUBTKAHgGcosYwBzAPgEpEBvAX0QHofEAwnByloAGzyJSOYVQB0AWABQy0JFgJEAI2yYc%2BIgCVSGACYIJZSjTqNWHRNz6IAkpGFZRUCVJlZ5ykA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIVrEB0AQFtE%2BXrSKlUGSL2gjokcGAC%2BII0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function foo(array: Set<string>) {} // Correctly errors.
function bar(array: ReadonlySet<string>) { } // Incorrectly errors.
```

**Comment by bradzacher** [comment]
The reason that this errors is because whilst the underlying data store is immutable - the class itself is not.

But you can fix this by using the [`treatMethodsAsReadonly`](https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBDATljBPALkQGUBTKAHgGcosYwBzAPgEpEBvAX0QHofEAwnByloAGzyJSOYVQB0AWABQy0JFgJEAI2yYc%2BIgCVSGACYIJZSjTqNWHRNz6IAkpGFZRUCVJlZ5ykA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIVrEB0AQFtE%2BXrSKlUGANrhsOXtBHRIAGm06sO7JHyD8AWUUALEUOQBBZACVEwsRPRhrWERTbABfEIBdbXDwoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) option.

Note that we don't seek to understand the purity of methods as TS doesn't track it - hence both `Set` and `ReadonlySet` are treated the same here.

**Comment by Zamiell** [comment]
In your program, the first line (i.e. the mutable set) no longer errors.
Thus, `treatMethodsAsReadonly` doesn't seem to solve this problem.
I propose that we should hard-code an exception for `ReadonlyMap` and `ReadonlySet` into the rule itself.

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/prefer-readonly-parameter-types#allow

The `allow` option will allow you to add a special case for this specific type.

Eg `"allow" : [{"from" : "lib", "name": "ReadonlySet"}]` ([playground](https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBDATljBPALkQGUBTKAHgGcosYwBzAPgEpEBvAX0QHofEAwnByloAGzyJSOYVQB0AWABQy0JFgJEAI2yYc%2BIgCVSGACYIJZSjTqNWHRNz6IAkpGFZRUCVJlZ5ykA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIVrEB0AQFtE%2BXrSKlUGANrhsOXtBHRIAGm06sO7JHyD8AWUUALEUOQBBZACVEwsRPRguAXhkRBMLSyD4EQB3fy1ws1MEnC59OX9IRgAjYyTkyCZ5RAyvH3FCAGVFSDyLAF9aiABdWoadFogGhqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

**Comment by Zamiell** [comment]
> We purposely don't want to encode assumptions about things into the rule because each codebase is different and has different constraints they want to enforce.

I totally get that, but don't you think that treating `ReadonlyMap` as equivalent to `Readonly<Map>` is pretty safe assumption, all things considered? (There is already a hardcoded exception in the code for this rule to allow `Readonly<T>`, so I am essentially just suggesting that this hard-coded exception be expanded to `ReadonlyMap` and `ReadonlySet`.)

I'd be really, really surprised if there was a codebase out there that wanted to allow `Readonly<Map>` but not allow `ReadonlyMap`, that seems super bizarre and not something we should support.
